---
title: "Nucleotide Diversity"
author: "Madeline Carpenter"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir="~/projects/eco_genomics_2025/population_genomics/myresults/ANGSD/PCA_ADMIX/")
```

# Import Libraries

```{r}
library(ggplot2)
library(tidyverse)
```

# Read in sample names

```{r}
names <- read.table("/gpfs1/home/m/c/mcarpen3/projects/eco_genomics_2025/population_genomics/myresults/ANGSD/PCA_ADMIX/RSBS_bam.list")
names <- unlist(strsplit(basename(as.character(names[,1])), split = ".sorted.rmdup.bam"))
split = strsplit(names, "_")
pops <- data.frame(names[1:113], do.call(rbind, split[1:113]))
names(pops) = c("Ind", "Pop", "Row", "Col")

## rank order the samples according to population code
ord <- order(pops[,2])
```

# Examine covariance matrix

```{r}
kit <- c(2:5)

for(i in 1:4){
  kval <- kit[i]
  
  cov <- as.matrix(read.table(paste0("/gpfs1/home/m/c/mcarpen3/projects/eco_genomics_2025/population_genomics/myresults/ANGSD/PCA_ADMIX/all_RS_K", kval, "_hw.cov")))

  pca <- eigen(cov)

  var <- round(pca$value/sum(pca$values),3)

  var[1:3]

  barplot(var, xlab=paste0("Eigenvalues of the PCA at K=", kval),
        ylab="Proportion of variance explained")

}

```

# Read in ADMIX data

```{r}
kit <- c(2:5)

for(i in 1:4){
  kval <- kit[i]
  kname <- paste0("adK_", kval)
  kname <- read.table(paste0("/gpfs1/home/m/c/mcarpen3/projects/eco_genomics_2025/population_genomics/myresults/ANGSD/PCA_ADMIX/all_RS_K", kval, "_hw.admix.", kval, ".Q"), sep=" ", header=F)
  K=dim(kname)[2] #Find the level of K modeled
  
  barplot(t(kname)[,ord],
        space=0,border=NA,
        xlab="Populations",ylab="Admixture proportions",
        main=paste0("Red Spruce-Black Spruce Admixture: K=",K))
text(tapply(1:nrow(pops),pops[ord,2],mean),-0.05,unique(pops[ord,2]),xpd=T,cex=0.75,srt=45)
abline(v=cumsum(sapply(unique(pops[ord,2]),function(x){sum(pops[ord,2]==x)})),col=1,lwd=1.2)
}
```
